###  Task- Performing Data Cleaning, and analysis on "Titanic" dataset

In [31]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier

In [7]:
titanic=pd.read_csv("TitanicTrain.csv") 
#Loading dataset from the csv file into "Pandas Dataframe"
titanic.head() #Head shows the first 5 entries of the dataframe

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

## Performing Data Cleaning and Analysis
#### 1. Understanding meaning of each column:
<br>Data Dictionary:
<br>**Variable        Description**</br>
1. Survived	- Survived (1) or died (0)
2. Pclass -	Passenger’s class (1 = 1st, 2 = 2nd, 3 = 3rd)
3. Name	- Passenger’s name
4. Sex -	Passenger’s sex
5. Age	- Passenger’s age
6. SibSp -	Number of siblings/spouses aboard
7. Parch -	Number of parents/children aboard (Some children travelled only with a nanny, therefore parch=0 for them.)
8. Ticket -	Ticket number
9. Fare -	Fare
10. Cabin -	Cabin
11. Embarked -	Port of embarkation (C = Cherbourg, Q = Queenstown, S = Southampton)

#### 2. Analysing which columns are completely useless in predicting the survival and deleting them
**Note** - Don't just delete the columns because you are not finding it useful. Or focus is not on deleting the columns. Our focus is on analysing how each column is affecting the result or the prediction and in accordance with that deciding whether to keep the column or to delete the column or fill the null values of the column by some values and if yes, then what values.



In [8]:
#Name column can never decide survival of a person, hence we can safely delete it
#Better to display the entire dataframe after a deletion of a column to get a better view
del titanic["Name"]
titanic.head()

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch  \
0            1         0       3    male  22.0      1      0   
1            2         1       1  female  38.0      1      0   
2            3         1       3  female  26.0      0      0   
3            4         1       1  female  35.0      1      0   
4            5         0       3    male  35.0      0      0   

             Ticket     Fare Cabin Embarked  
0         A/5 21171   7.2500   NaN        S  
1          PC 17599  71.2833   C85        C  
2  STON/O2. 3101282   7.9250   NaN        S  
3            113803  53.1000  C123        S  
4            373450   8.0500   NaN        S

In [9]:
del titanic["PassengerId"]
titanic.head()

Survived  Pclass     Sex   Age  SibSp  Parch            Ticket     Fare  \
0         0       3    male  22.0      1      0         A/5 21171   7.2500   
1         1       1  female  38.0      1      0          PC 17599  71.2833   
2         1       3  female  26.0      0      0  STON/O2. 3101282   7.9250   
3         1       1  female  35.0      1      0            113803  53.1000   
4         0       3    male  35.0      0      0            373450   8.0500   

  Cabin Embarked  
0   NaN        S  
1   C85        C  
2   NaN        S  
3  C123        S  
4   NaN        S

In [10]:
del titanic["Ticket"]
titanic.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Cabin Embarked
0         0       3    male  22.0      1      0   7.2500   NaN        S
1         1       1  female  38.0      1      0  71.2833   C85        C
2         1       3  female  26.0      0      0   7.9250   NaN        S
3         1       1  female  35.0      1      0  53.1000  C123        S
4         0       3    male  35.0      0      0   8.0500   NaN        S

In [11]:
#To see how many entries are there



titanic.shape

(891, 9)

In [12]:
#To see how many null entries are there in each column
#Here isnull() returns 1 if there is a null entry adding upon 
#which(using sum() function) we get number of null entries in each column 



titanic.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Cabin       687
Embarked      2
dtype: int64

In [13]:
del titanic["Cabin"]
titanic.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S
3         1       1  female  35.0      1      0  53.1000        S
4         0       3    male  35.0      0      0   8.0500        S

In [14]:
#Applying statistical approach on the above dataframe to analyse 
#which feature or column is affecting the survival rate and which is useless colum
titanic.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [15]:
del titanic["Fare"]
titanic.head()

Survived  Pclass     Sex   Age  SibSp  Parch Embarked
0         0       3    male  22.0      1      0        S
1         1       1  female  38.0      1      0        C
2         1       3  female  26.0      0      0        S
3         1       1  female  35.0      1      0        S
4         0       3    male  35.0      0      0        S

#### We want to check if "Embarked" column is is important for analysis or not, that is whether survival of the person depends on the Embarked column value or not. 

In [16]:
# Finding the number of people who have survived 
# given that they have embarked or boarded from a particular port

survivedS = titanic[titanic.Embarked == 'Q'][titanic.Survived == 1].shape[0]
survivedS

#similarly checked for C and S too and the results were
# Survived for C=93
# Survived for S=217

G:\anaconda\lib\site-packages\ipykernel_launcher.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  after removing the cwd from sys.path.


30

In [17]:
# Finding the number of people who have not survived 
# given that they have embarked or boarded from a particular port


survivedNS= titanic[titanic.Embarked=="Q"][titanic.Survived==0].shape[0]
survivedNS

#similarly checked for C and S too and the results were
# Not Survived for C = 75
# Not Survived for S = 427

G:\anaconda\lib\site-packages\ipykernel_launcher.py:5: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """


47

As there are significant changes in the survival rate based on which port the passengers aboard the ship.
<br>
We cannot delete the whole embarked column(It is useful).
<br>
Now the Embarked column has two null values in it with total entries 891 rows and hence we can safely say that deleting two rows from 891 rows will not affect the result. So rather than trying to fill those null values with some vales. We can simply remove them.
<br>
For the Null values in Embarked we can remove the rows with null values
<br>
perform **dropna()** on the Datafrem when only these values are present as null
<br>
titanic.dropna(inplace = True)
<br>
Note - inplace = True , so that the changes takes place in the same dataframe and not in the copy
<br>
titanic.shape

#### Changing Value for "Male, Female" string values to numeric values , male=1 and female=2

In [18]:
def getNumber(str):
    if str=="male":
        return 1
    else:
        return 2
titanic["gender"]=titanic["Sex"].apply(getNumber)
#We have created a new column called "gender" and 
#filling it with values 1 ,2 based on the values of sex column
titanic.head()

Survived  Pclass     Sex   Age  SibSp  Parch Embarked  gender
0         0       3    male  22.0      1      0        S       1
1         1       1  female  38.0      1      0        C       2
2         1       3  female  26.0      0      0        S       2
3         1       1  female  35.0      1      0        S       2
4         0       3    male  35.0      0      0        S       1

In [19]:
#Deleting Sex column, since no use of it now
del titanic["Sex"]
titanic.head()

Survived  Pclass   Age  SibSp  Parch Embarked  gender
0         0       3  22.0      1      0        S       1
1         1       1  38.0      1      0        C       2
2         1       3  26.0      0      0        S       2
3         1       1  35.0      1      0        S       2
4         0       3  35.0      0      0        S       1

#### Drawing a pie chart for number of males and females aboard

In [20]:
import matplotlib.pyplot as plt
from matplotlib import style

males = (titanic['gender'] == 1).sum() 
#Summing up all the values of column gender with a 
#condition for male and similary for females
females = (titanic['gender'] == 2).sum()
print(males)
print(females)
p = [males, females]
plt.pie(p,    #giving array
       labels = ['Male', 'Female'], #Correspndingly giving labels
       colors = ['blue', 'red'],   # Corresponding colors
       explode = (0.15, 0),    #How much the gap should me there between the pies
       startangle = 0)  #what start angle should be given
plt.axis('equal') 
plt.show()

577
314


#### Fill the null values of the Age column. Fill mean Survived age(mean age of the survived people) in the column where the person has survived and mean not Survived age (mean age of the people who have not survived) in the column where person has not survived

In [21]:
#finding mean survived age
meanS= titanic[titanic.Survived==1].Age.mean()
meanS

28.343689655172415

#### Creating a new "age" column , filling values in it with a condition (given inside) if goes True then given values (here meanS) is put in place of last values else nothing happens, simply the values are copied from the "Age" column of the daatset  

In [22]:
titanic["age"]=np.where(pd.isnull(titanic.Age) & titanic["Survived"]==1  ,meanS, titanic["Age"])
titanic

Survived  Pclass   Age  SibSp  Parch Embarked  gender       age
0           0       3  22.0      1      0        S       1  22.00000
1           1       1  38.0      1      0        C       2  38.00000
2           1       3  26.0      0      0        S       2  26.00000
3           1       1  35.0      1      0        S       2  35.00000
4           0       3  35.0      0      0        S       1  35.00000
5           0       3   NaN      0      0        Q       1       NaN
6           0       1  54.0      0      0        S       1  54.00000
7           0       3   2.0      3      1        S       1   2.00000
8           1       3  27.0      0      2        S       2  27.00000
9           1       2  14.0      1      0        C       2  14.00000
10          1       3   4.0      1      1        S       2   4.00000
11          1       1  58.0      0      0        S       2  58.00000
12          0       3  20.0      0      0        S       1  20.00000
13          0       3  39.0      1      5        S       1  39.00000
14          0       3  14.0      0      0        S       2  14.00000
15          1       2  55.0      0      0        S       2  55.00000
16          0       3   2.0      4      1        Q       1   2.00000
17          1       2   NaN      0      0        S       1  28.34369
18          0       3  31.0      1      0        S       2  31.00000
19          1       3   NaN      0      0        C       2  28.34369
20          0       2  35.0      0      0        S       1  35.00000
21          1       2  34.0      0      0        S       1  34.00000
22          1       3  15.0      0      0        Q       2  15.00000
23          1       1  28.0      0      0        S       1  28.00000
24          0       3   8.0      3      1        S       2   8.00000
25          1       3  38.0      1      5        S       2  38.00000
26          0       3   NaN      0      0        C       1       NaN
27          0       1  19.0      3      2        S       1  19.00000
28          1       3   NaN      0      0        Q       2  28.34369
29          0       3   NaN      0      0        S       1       NaN
..        ...     ...   ...    ...    ...      ...     ...       ...
861         0       2  21.0      1      0        S       1  21.00000
862         1       1  48.0      0      0        S       2  48.00000
863         0       3   NaN      8      2        S       2       NaN
864         0       2  24.0      0      0        S       1  24.00000
865         1       2  42.0      0      0        S       2  42.00000
866         1       2  27.0      1      0        C       2  27.00000
867         0       1  31.0      0      0        S       1  31.00000
868         0       3   NaN      0      0        S       1       NaN
869         1       3   4.0      1      1        S       1   4.00000
870         0       3  26.0      0      0        S       1  26.00000
871         1       1  47.0      1      1        S       2  47.00000
872         0       1  33.0      0      0        S       1  33.00000
873         0       3  47.0      0      0        S       1  47.00000
874         1       2  28.0      1      0        C       2  28.00000
875         1       3  15.0      0      0        C       2  15.00000
876         0       3  20.0      0      0        S       1  20.00000
877         0       3  19.0      0      0        S       1  19.00000
878         0       3   NaN      0      0        S       1       NaN
879         1       1  56.0      0      1        C       2  56.00000
880         1       2  25.0      0      1        S       2  25.00000
881         0       3  33.0      0      0        S       1  33.00000
882         0       3  22.0      0      0        S       2  22.00000
883         0       2  28.0      0      0        S       1  28.00000
884         0       3  25.0      0      0        S       1  25.00000
885         0       3  39.0      0      5        Q       2  39.00000
886         0       2  27.0      0      0        S       1  27.00000
887

In [23]:
# Finding the mean age of "Not Survived" people
meanNS=titanic[titanic.Survived==0].Age.mean()
meanNS

30.62617924528302

Now the "age" column contains null entries in the places where survived is equal to zero.
<br>
Filling those null values in one go with meanNS 

In [24]:
titanic.age.fillna(meanNS,inplace=True)
titanic

Survived  Pclass   Age  SibSp  Parch Embarked  gender        age
0           0       3  22.0      1      0        S       1  22.000000
1           1       1  38.0      1      0        C       2  38.000000
2           1       3  26.0      0      0        S       2  26.000000
3           1       1  35.0      1      0        S       2  35.000000
4           0       3  35.0      0      0        S       1  35.000000
5           0       3   NaN      0      0        Q       1  30.626179
6           0       1  54.0      0      0        S       1  54.000000
7           0       3   2.0      3      1        S       1   2.000000
8           1       3  27.0      0      2        S       2  27.000000
9           1       2  14.0      1      0        C       2  14.000000
10          1       3   4.0      1      1        S       2   4.000000
11          1       1  58.0      0      0        S       2  58.000000
12          0       3  20.0      0      0        S       1  20.000000
13          0       3  39.0      1      5        S       1  39.000000
14          0       3  14.0      0      0        S       2  14.000000
15          1       2  55.0      0      0        S       2  55.000000
16          0       3   2.0      4      1        Q       1   2.000000
17          1       2   NaN      0      0        S       1  28.343690
18          0       3  31.0      1      0        S       2  31.000000
19          1       3   NaN      0      0        C       2  28.343690
20          0       2  35.0      0      0        S       1  35.000000
21          1       2  34.0      0      0        S       1  34.000000
22          1       3  15.0      0      0        Q       2  15.000000
23          1       1  28.0      0      0        S       1  28.000000
24          0       3   8.0      3      1        S       2   8.000000
25          1       3  38.0      1      5        S       2  38.000000
26          0       3   NaN      0      0        C       1  30.626179
27          0       1  19.0      3      2        S       1  19.000000
28          1       3   NaN      0      0        Q       2  28.343690
29          0       3   NaN      0      0        S       1  30.626179
..        ...     ...   ...    ...    ...      ...     ...        ...
861         0       2  21.0      1      0        S       1  21.000000
862         1       1  48.0      0      0        S       2  48.000000
863         0       3   NaN      8      2        S       2  30.626179
864         0       2  24.0      0      0        S       1  24.000000
865         1       2  42.0      0      0        S       2  42.000000
866         1       2  27.0      1      0        C       2  27.000000
867         0       1  31.0      0      0        S       1  31.000000
868         0       3   NaN      0      0        S       1  30.626179
869         1       3   4.0      1      1        S       1   4.000000
870         0       3  26.0      0      0        S       1  26.000000
871         1       1  47.0      1      1        S       2  47.000000
872         0       1  33.0      0      0        S       1  33.000000
873         0       3  47.0      0      0        S       1  47.000000
874         1       2  28.0      1      0        C       2  28.000000
875         1       3  15.0      0      0        C       2  15.000000
876         0       3  20.0      0      0        S       1  20.000000
877         0       3  19.0      0      0        S       1  19.000000
878         0       3   NaN      0      0        S       1  30.626179
879         1       1  56.0      0      1        C       2  56.000000
880         1       2  25.0      0      1        S       2  25.000000
881         0       3  33.0      0      0        S       1  33.000000
882         0       3  22.0      0      0        S       2  22.000000
883         0       2  28.0      0      0        S       1  28.000000
884         0       3  25.0      0      0        S       1  25.000000
885         0       3  39.0      0      5        Q       2  39.000000
886         0  

In [25]:
#We can safely delete the Age column
del titanic["Age"]
titanic

Survived  Pclass  SibSp  Parch Embarked  gender        age
0           0       3      1      0        S       1  22.000000
1           1       1      1      0        C       2  38.000000
2           1       3      0      0        S       2  26.000000
3           1       1      1      0        S       2  35.000000
4           0       3      0      0        S       1  35.000000
5           0       3      0      0        Q       1  30.626179
6           0       1      0      0        S       1  54.000000
7           0       3      3      1        S       1   2.000000
8           1       3      0      2        S       2  27.000000
9           1       2      1      0        C       2  14.000000
10          1       3      1      1        S       2   4.000000
11          1       1      0      0        S       2  58.000000
12          0       3      0      0        S       1  20.000000
13          0       3      1      5        S       1  39.000000
14          0       3      0      0        S       2  14.000000
15          1       2      0      0        S       2  55.000000
16          0       3      4      1        Q       1   2.000000
17          1       2      0      0        S       1  28.343690
18          0       3      1      0        S       2  31.000000
19          1       3      0      0        C       2  28.343690
20          0       2      0      0        S       1  35.000000
21          1       2      0      0        S       1  34.000000
22          1       3      0      0        Q       2  15.000000
23          1       1      0      0        S       1  28.000000
24          0       3      3      1        S       2   8.000000
25          1       3      1      5        S       2  38.000000
26          0       3      0      0        C       1  30.626179
27          0       1      3      2        S       1  19.000000
28          1       3      0      0        Q       2  28.343690
29          0       3      0      0        S       1  30.626179
..        ...     ...    ...    ...      ...     ...        ...
861         0       2      1      0        S       1  21.000000
862         1       1      0      0        S       2  48.000000
863         0       3      8      2        S       2  30.626179
864         0       2      0      0        S       1  24.000000
865         1       2      0      0        S       2  42.000000
866         1       2      1      0        C       2  27.000000
867         0       1      0      0        S       1  31.000000
868         0       3      0      0        S       1  30.626179
869         1       3      1      1        S       1   4.000000
870         0       3      0      0        S       1  26.000000
871         1       1      1      1        S       2  47.000000
872         0       1      0      0        S       1  33.000000
873         0       3      0      0        S       1  47.000000
874         1       2      1      0        C       2  28.000000
875         1       3      0      0        C       2  15.000000
876         0       3      0      0        S       1  20.000000
877         0       3      0      0        S       1  19.000000
878         0       3      0      0        S       1  30.626179
879         1       1      0      1        C       2  56.000000
880         1       2      0      1        S       2  25.000000
881         0       3      0      0        S       1  33.000000
882         0       3      0      0        S       2  22.000000
883         0       2      0      0        S       1  28.000000
884         0       3      0      0        S       1  25.000000
885         0       3      0      5        Q       2  39.000000
886         0       2      0      0        S       1  27.000000
887         1       1      0      0        S       2  19.000000
888         0       3      1      2        S       2  30.626179
889         1       1      0      0        C       1  26.000000
890         0       3      0      0        Q       1  32.000000

[891 rows x 7 columns]

In [26]:
#Renaming "age" and "gender" columns
titanic.rename(columns={'age':'Age'}, inplace=True)
titanic.rename(columns={'gender':'Sex'}, inplace=True)
titanic.head()


Survived  Pclass  SibSp  Parch Embarked  Sex   Age
0         0       3      1      0        S    1  22.0
1         1       1      1      0        C    2  38.0
2         1       3      0      0        S    2  26.0
3         1       1      1      0        S    2  35.0
4         0       3      0      0        S    1  35.0

#### Now removing two rows where embarked value was null (Discussed in the beginning)

In [27]:
titanic.dropna(inplace=True)
titanic.head()

Survived  Pclass  SibSp  Parch Embarked  Sex   Age
0         0       3      1      0        S    1  22.0
1         1       1      1      0        C    2  38.0
2         1       3      0      0        S    2  26.0
3         1       1      1      0        S    2  35.0
4         0       3      0      0        S    1  35.0

In [28]:
del titanic["Embarked"]

In [29]:
titanic.head()

Survived  Pclass  SibSp  Parch  Sex   Age
0         0       3      1      0    1  22.0
1         1       1      1      0    2  38.0
2         1       3      0      0    2  26.0
3         1       1      1      0    2  35.0
4         0       3      0      0    1  35.0

In [32]:
clf=KNeighborsClassifier()

In [33]:
X_train,X_test,y_train,y_test=train_test_split(titanic,titanic.Survived)

In [34]:
clf.fit(X_train,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')

In [36]:
print(clf.score(X_train,y_train))
print(clf.score(X_test,y_test))

0.9654654654654654
0.9237668161434978


In [37]:

log_reg=LogisticRegression(C=1)

log_reg.fit(X_train,y_train)

LogisticRegression(C=1, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [38]:
print(log_reg.score(X_train,y_train))
print(log_reg.score(X_test,y_test))

1.0
1.0
